nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -91,23 +182,47 @@ impl DataStore {
         //     .authorize(authz::Action::ListChildren, &authz::IP_POOL_LIST)
         //     .await?;
 
-        dsl::ip_pool
-            .filter(dsl::silo_id.eq(authz_silo_id).or(dsl::silo_id.is_null()))
-            .filter(dsl::is_default.eq(true))
-            .filter(dsl::time_deleted.is_null())
-            // this will sort by most specific first, i.e.,
-            //
-            //   (silo)
-            //   (null)
-            //
-            // then by only taking the first result, we get the most specific one
-            .order(dsl::silo_id.asc().nulls_last())
+        // join ip_pool to ip_pool_resource and filter
+
+        // used in both success and error outcomes
+        let lookup_type = LookupType::ByCompositeId(
+            "Default pool for current silo".to_string(),
+        );
+
+        ip_pool::table
+            .inner_join(ip_pool_resource::table)
+            .filter(
+                ip_pool_resource::resource_type.eq(IpPoolResourceType::Silo),
+            )
+            .filter(ip_pool_resource::resource_id.eq(authz_silo_id))
+            .filter(ip_pool_resource::is_default.eq(true))
+            .filter(ip_pool::time_deleted.is_null())
+            // Order by most specific first so we get the most specific.
+            // resource_type is an enum in the DB and therefore gets its order
+            // from the definition; it's not lexicographic. So correctness here
+            // relies on the types being most-specific-first in the definition.
+            // There are tests for this.
+            .order(ip_pool_resource::resource_type.asc())
             .select(IpPool::as_select())
             .first_async::<IpPool>(
                 &*self.pool_connection_authorized(opctx).await?,
             )
             .await
-            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+            .map_err(|e| {
+                // janky to do this manually, but this is an unusual kind of
+                // lookup in that it is by (silo_id, is_default=true), which is
+                // arguably a composite ID.
+                public_error_from_diesel_lookup(
+                    e,
+                    ResourceType::IpPool,
+                    &lookup_type,
+                )
+            })
+            .map(|ip_pool| {
+                let authz_pool =
+                    authz::IpPool::new(authz::FLEET, ip_pool.id(), lookup_type);
+                (authz_pool, ip_pool)
```

Comment:
This is interesting, I haven't noticed where we've done something like this explicitly in the data store before. I'm not sure why the authz / resource tuple was necessary.

Reply:
Ahh, I found it below. This is a non-standard lookup so the id/name generated lookups that hang off of nexus wouldn't be available and thus its existence.

Reply:
Right. An edge case that doesn't really fit in the look up by name, id, or composite id model. Arguably it's a composite ID, which is why I'm using it as a sort of catchall for this.